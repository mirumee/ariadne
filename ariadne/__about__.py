__version__ = "0.0.1.dev0"  # This is overwritten by Hatch in CI/CD, don't change it.
