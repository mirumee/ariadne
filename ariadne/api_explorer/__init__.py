from .api_explorer import APIExplorer, APIExplorerHttp405
from .graphiql import APIExplorerGraphiQL
from .playground import APIExplorerPlayground

__all__ = [
    "APIExplorer",
    "APIExplorerGraphiQL",
    "APIExplorerHttp405",
    "APIExplorerPlayground",
]
