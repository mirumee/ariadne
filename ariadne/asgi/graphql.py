from typing import Optional

from graphql import GraphQLSchema
from starlette.types import Receive, Scope, Send

from ..format_error import format_error
from ..types import (
    ContextValue,
    ErrorFormatter,
    RootValue,
    ValidationRules,
)
from .handlers import (
    GraphQLHTTPHandler,
    GraphQLWebsocketHandler,
    GraphQLWSHandler,
)


class GraphQL:
    def __init__(
        self,
        schema: GraphQLSchema,
        *,
        context_value: Optional[ContextValue] = None,
        root_value: Optional[RootValue] = None,
        validation_rules: Optional[ValidationRules] = None,
        debug: bool = False,
        introspection: bool = True,
        logger: Optional[str] = None,
        error_formatter: ErrorFormatter = format_error,
        http_handler: Optional[GraphQLHTTPHandler] = None,
        websocket_handler: Optional[GraphQLWebsocketHandler] = None,
    ):
        if http_handler:
            self.http_handler = http_handler
        else:
            self.http_handler = GraphQLHTTPHandler()

        if websocket_handler:
            self.websocket_handler = websocket_handler
        else:
            self.websocket_handler = GraphQLWSHandler()

        self.http_handler.configure(
            schema,
            context_value,
            root_value,
            validation_rules,
            debug,
            introspection,
            logger,
            error_formatter,
        )
        self.websocket_handler.configure(
            schema,
            context_value,
            root_value,
            validation_rules,
            debug,
            introspection,
            logger,
            error_formatter,
            http_handler=self.http_handler,
        )

    async def __call__(self, scope: Scope, receive: Receive, send: Send):
        if scope["type"] == "http":
            await self.http_handler.handle(scope=scope, receive=receive, send=send)
        elif scope["type"] == "websocket":
            await self.websocket_handler.handle(scope=scope, receive=receive, send=send)
        else:
            raise ValueError("Unknown scope type: %r" % (scope["type"],))
