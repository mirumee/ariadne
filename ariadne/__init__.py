from .enums import (
    EnumType,
    set_default_enum_values_on_schema,
    validate_schema_enum_values,
)
from .executable_schema import make_executable_schema
from .extensions import ExtensionManager
from .file_uploads import combine_multipart_data, upload_scalar
from .format_error import (
    format_error,
    get_error_extension,
    get_formatted_error_context,
    get_formatted_error_traceback,
)
from .graphql import graphql, graphql_sync, subscribe
from .inputs import InputType
from .interfaces import InterfaceType, type_implements_interface
from .load_schema import load_schema_from_path
from .objects import MutationType, ObjectType, QueryType
from .resolvers import (
    FallbackResolversSetter,
    SnakeCaseFallbackResolversSetter,
    fallback_resolvers,
    is_default_resolver,
    resolve_to,
    snake_case_fallback_resolvers,
)
from .scalars import ScalarType
from .schema_names import SchemaNameConverter, convert_schema_names
from .schema_visitor import SchemaDirectiveVisitor
from .subscriptions import SubscriptionType
from .types import Extension, SchemaBindable
from .unions import UnionType
from .utils import (
    convert_camel_case_to_snake,
    convert_kwargs_to_snake_case,
    gql,
    unwrap_graphql_error,
)

__all__ = [
    "EnumType",
    "Extension",
    "ExtensionManager",
    "FallbackResolversSetter",
    "InputType",
    "InterfaceType",
    "MutationType",
    "ObjectType",
    "QueryType",
    "ScalarType",
    "SchemaBindable",
    "SchemaDirectiveVisitor",
    "SchemaNameConverter",
    "SnakeCaseFallbackResolversSetter",
    "SubscriptionType",
    "UnionType",
    "combine_multipart_data",
    "convert_camel_case_to_snake",
    "convert_kwargs_to_snake_case",
    "convert_schema_names",
    "fallback_resolvers",
    "format_error",
    "get_error_extension",
    "get_formatted_error_context",
    "get_formatted_error_traceback",
    "gql",
    "graphql",
    "graphql_sync",
    "is_default_resolver",
    "load_schema_from_path",
    "make_executable_schema",
    "resolve_to",
    "set_default_enum_values_on_schema",
    "snake_case_fallback_resolvers",
    "subscribe",
    "type_implements_interface",
    "unwrap_graphql_error",
    "upload_scalar",
    "validate_schema_enum_values",
]
