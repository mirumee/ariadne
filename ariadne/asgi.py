import asyncio
import json
from typing import (
    Any,
    AsyncGenerator,
    Callable,
    Dict,
    List,
    Optional,
    Union,
    Type,
    cast,
)

from graphql import GraphQLError, GraphQLSchema
from graphql.execution import MiddlewareManager
from starlette.datastructures import UploadFile
from starlette.requests import Request
from starlette.responses import HTMLResponse, JSONResponse, PlainTextResponse, Response
from starlette.types import Receive, Scope, Send
from starlette.websockets import WebSocket, WebSocketState, WebSocketDisconnect

from .constants import DATA_TYPE_JSON, DATA_TYPE_MULTIPART, PLAYGROUND_HTML
from .exceptions import HttpBadRequestError, HttpError
from .file_uploads import combine_multipart_data
from .format_error import format_error
from .graphql import graphql, subscribe
from .logger import log_error
from .types import ContextValue, ErrorFormatter, Extension, RootValue

GQL_CONNECTION_INIT = "connection_init"  # Client -> Server
GQL_CONNECTION_ACK = "connection_ack"  # Server -> Client
GQL_CONNECTION_ERROR = "connection_error"  # Server -> Client

# NOTE: The keep alive message type does not follow the standard due to connection optimizations
GQL_CONNECTION_KEEP_ALIVE = "ka"  # Server -> Client

GQL_CONNECTION_TERMINATE = "connection_terminate"  # Client -> Server
GQL_START = "start"  # Client -> Server
GQL_DATA = "data"  # Server -> Client
GQL_ERROR = "error"  # Server -> Client
GQL_COMPLETE = "complete"  # Server -> Client
GQL_STOP = "stop"  # Client -> Server

ExtensionList = Optional[List[Type[Extension]]]


class GraphQL:
    def __init__(
        self,
        schema: GraphQLSchema,
        *,
        context_value: Optional[ContextValue] = None,
        root_value: Optional[RootValue] = None,
        debug: bool = False,
        logger: Optional[str] = None,
        error_formatter: ErrorFormatter = format_error,
        extensions: Union[Callable[[Any], ExtensionList], ExtensionList] = None,
        middleware: Optional[MiddlewareManager] = None,
        keepalive: float = None,
    ):
        self.context_value = context_value
        self.root_value = root_value
        self.debug = debug
        self.logger = logger
        self.error_formatter = error_formatter
        self.extensions = extensions
        self.middleware = middleware
        self.keepalive = keepalive
        self.schema = schema

    async def __call__(self, scope: Scope, receive: Receive, send: Send):
        if scope["type"] == "http":
            await self.handle_http(scope=scope, receive=receive, send=send)
        elif scope["type"] == "websocket":
            await self.handle_websocket(scope=scope, receive=receive, send=send)
        else:
            raise ValueError("Unknown scope type: %r" % (scope["type"],))

    async def get_context_for_request(self, request: Any) -> Any:
        if callable(self.context_value):
            return self.context_value(request)
        return self.context_value or {"request": request}

    async def get_extensions_for_request(self, request: Any) -> ExtensionList:
        if callable(self.extensions):
            return self.extensions(request)
        return self.extensions

    async def handle_http(self, scope: Scope, receive: Receive, send: Send):
        request = Request(scope=scope, receive=receive)
        if request.method == "GET":
            response = await self.render_playground(request)
        elif request.method == "POST":
            response = await self.graphql_http_server(request)
        else:
            response = Response(status_code=405)
        await response(scope, receive, send)

    async def handle_websocket(self, scope: Scope, receive: Receive, send: Send):
        websocket = WebSocket(scope=scope, receive=receive, send=send)
        await self.websocket_server(websocket)

    async def render_playground(  # pylint: disable=unused-argument
        self, request: Request
    ) -> Response:
        return HTMLResponse(PLAYGROUND_HTML)

    async def graphql_http_server(self, request: Request) -> Response:
        try:
            data = await self.extract_data_from_request(request)
        except HttpError as error:
            return PlainTextResponse(error.message or error.status, status_code=400)

        context_value = await self.get_context_for_request(request)
        extensions = await self.get_extensions_for_request(request)

        success, response = await graphql(
            self.schema,
            data,
            context_value=context_value,
            root_value=self.root_value,
            debug=self.debug,
            logger=self.logger,
            error_formatter=self.error_formatter,
            extensions=extensions,
            middleware=self.middleware,
        )
        status_code = 200 if success else 400
        return JSONResponse(response, status_code=status_code)

    async def extract_data_from_request(self, request: Request):
        content_type = request.headers.get("Content-Type", "")
        content_type = content_type.split(";")[0]

        if content_type == DATA_TYPE_JSON:
            return await self.extract_data_from_json_request(request)
        if content_type == DATA_TYPE_MULTIPART:
            return await self.extract_data_from_multipart_request(request)

        raise HttpBadRequestError(
            "Posted content must be of type {} or {}".format(
                DATA_TYPE_JSON, DATA_TYPE_MULTIPART
            )
        )

    async def extract_data_from_json_request(self, request: Request):
        try:
            return await request.json()
        except (TypeError, ValueError):
            raise HttpBadRequestError("Request body is not a valid JSON")

    async def extract_data_from_multipart_request(self, request: Request):
        try:
            request_body = await request.form()
        except ValueError:
            raise HttpBadRequestError("Request body is not a valid multipart/form-data")

        try:
            operations = json.loads(request_body.get("operations"))
        except (TypeError, ValueError):
            raise HttpBadRequestError(
                "Request 'operations' multipart field is not a valid JSON"
            )
        try:
            files_map = json.loads(request_body.get("map"))
        except (TypeError, ValueError):
            raise HttpBadRequestError(
                "Request 'map' multipart field is not a valid JSON"
            )

        request_files = {
            key: value
            for key, value in request_body.items()
            if isinstance(value, UploadFile)
        }

        return combine_multipart_data(operations, files_map, request_files)

    async def websocket_server(self, websocket: WebSocket) -> None:
        subscriptions: Dict[str, AsyncGenerator] = {}
        await websocket.accept("graphql-ws")
        try:
            while websocket.application_state != WebSocketState.DISCONNECTED:
                message = await websocket.receive_json()
                await self.handle_websocket_message(message, websocket, subscriptions)
        except WebSocketDisconnect:
            pass
        finally:
            for operation_id in subscriptions:
                await subscriptions[operation_id].aclose()

    async def handle_websocket_message(
        self,
        message: dict,
        websocket: WebSocket,
        subscriptions: Dict[str, AsyncGenerator],
    ):
        operation_id = cast(str, message.get("id"))
        message_type = cast(str, message.get("type"))

        if message_type == GQL_CONNECTION_INIT:
            await websocket.send_json({"type": GQL_CONNECTION_ACK})
            asyncio.ensure_future(self.keep_websocket_alive(websocket))
        elif message_type == GQL_CONNECTION_TERMINATE:
            await websocket.close()
        elif message_type == GQL_START:
            await self.start_websocket_subscription(
                message.get("payload"), operation_id, websocket, subscriptions
            )
        elif message_type == GQL_STOP:
            if operation_id in subscriptions:
                await subscriptions[operation_id].aclose()
                del subscriptions[operation_id]

    async def keep_websocket_alive(self, websocket: WebSocket):
        if not self.keepalive:
            return
        while websocket.application_state != WebSocketState.DISCONNECTED:
            try:
                await websocket.send_json({"type": GQL_CONNECTION_KEEP_ALIVE})
            except WebSocketDisconnect:
                return
            await asyncio.sleep(self.keepalive)

    async def start_websocket_subscription(
        self,
        data: Any,
        operation_id: str,
        websocket: WebSocket,
        subscriptions: Dict[str, AsyncGenerator],
    ):
        context_value = await self.get_context_for_request(websocket)
        success, results = await subscribe(
            self.schema,
            data,
            context_value=context_value,
            root_value=self.root_value,
            debug=self.debug,
            logger=self.logger,
            error_formatter=self.error_formatter,
        )
        if not success:
            results = cast(List[dict], results)
            await websocket.send_json(
                {"type": GQL_ERROR, "id": operation_id, "payload": results[0]}
            )
        else:
            results = cast(AsyncGenerator, results)
            subscriptions[operation_id] = results
            asyncio.ensure_future(
                self.observe_async_results(results, operation_id, websocket)
            )

    async def observe_async_results(
        self, results: AsyncGenerator, operation_id: str, websocket: WebSocket
    ) -> None:
        try:
            async for result in results:
                payload = {}
                if result.data:
                    payload["data"] = result.data
                if result.errors:
                    for error in result.errors:
                        log_error(error, self.logger)
                    payload["errors"] = [
                        self.error_formatter(error, self.debug)
                        for error in result.errors
                    ]
                await websocket.send_json(
                    {"type": GQL_DATA, "id": operation_id, "payload": payload}
                )
        except Exception as error:
            if not isinstance(error, GraphQLError):
                error = GraphQLError(str(error), original_error=error)
            log_error(error, self.logger)
            payload = {"errors": [self.error_formatter(error, self.debug)]}
            await websocket.send_json(
                {"type": GQL_DATA, "id": operation_id, "payload": payload}
            )

        if websocket.application_state != WebSocketState.DISCONNECTED:
            await websocket.send_json({"type": GQL_COMPLETE, "id": operation_id})
