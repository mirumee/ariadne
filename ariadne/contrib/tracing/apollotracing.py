from datetime import datetime
from inspect import isawaitable
from typing import Any, List, Optional, cast

from graphql import GraphQLResolveInfo

from ...types import ContextValue, Extension, Resolver
from .utils import format_path, should_trace

try:
    from time import perf_counter_ns
except ImportError:
    # Py 3.6 fallback
    from time import perf_counter

    NS_IN_SECOND = 1000000000

    def perf_counter_ns() -> int:
        return int(perf_counter() * NS_IN_SECOND)


TIMESTAMP_FORMAT = "%Y-%m-%dT%H:%M:%S.%fZ"


class ApolloTracingExtension(Extension):
    def __init__(self, trace_default_resolver: bool = False) -> None:
        self.trace_default_resolver = trace_default_resolver
        self.start_date: Optional[datetime] = None
        self.start_timestamp: int = 0
        self.resolvers: List[dict] = []

        self._totals = None

    def request_started(self, context: ContextValue):
        self.start_date = datetime.utcnow()
        self.start_timestamp = perf_counter_ns()

    async def resolve(
        self, next_: Resolver, obj: Any, info: GraphQLResolveInfo, **kwargs
    ):
        if not should_trace(info, self.trace_default_resolver):
            result = next_(obj, info, **kwargs)
            if isawaitable(result):
                result = await result
            return result

        start_timestamp = perf_counter_ns()
        record = {
            "path": format_path(info.path),
            "parentType": str(info.parent_type),
            "fieldName": info.field_name,
            "returnType": str(info.return_type),
            "startOffset": start_timestamp - cast(int, self.start_timestamp),
        }
        self.resolvers.append(record)
        try:
            result = next_(obj, info, **kwargs)
            if isawaitable(result):
                result = await result
            return result
        finally:
            end_timestamp = perf_counter_ns()
            record["duration"] = end_timestamp - start_timestamp

    def get_totals(self):
        if self._totals is None:
            self._totals = self._get_totals()
        return self._totals

    def _get_totals(self):
        return {
            "start": self.start_date,
            "end": datetime.utcnow(),
            "duration": perf_counter_ns() - self.start_timestamp,
            "resolvers": self.resolvers,
        }

    def format(self, context: ContextValue):
        totals = self.get_totals()

        return {
            "tracing": {
                "version": 1,
                "startTime": totals["start"].strftime(TIMESTAMP_FORMAT),
                "endTime": totals["end"].strftime(TIMESTAMP_FORMAT),
                "duration": totals["duration"],
                "execution": {"resolvers": totals["resolvers"]},
            }
        }


class ApolloTracingExtensionSync(ApolloTracingExtension):
    def resolve(
        self, next_: Resolver, obj: Any, info: GraphQLResolveInfo, **kwargs
    ):  # pylint: disable=invalid-overridden-method
        if not should_trace(info):
            result = next_(obj, info, **kwargs)
            return result

        start_timestamp = perf_counter_ns()
        record = {
            "path": format_path(info.path),
            "parentType": str(info.parent_type),
            "fieldName": info.field_name,
            "returnType": str(info.return_type),
            "startOffset": start_timestamp - cast(int, self.start_timestamp),
        }
        self.resolvers.append(record)
        try:
            result = next_(obj, info, **kwargs)
            return result
        finally:
            end_timestamp = perf_counter_ns()
            record["duration"] = end_timestamp - start_timestamp
