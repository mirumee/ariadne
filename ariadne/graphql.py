from asyncio import ensure_future
from inspect import isawaitable
from typing import (
    Any,
    AsyncGenerator,
    Awaitable,
    Collection,
    List,
    Optional,
    Sequence,
    Type,
    cast,
)

from graphql import (
    DocumentNode,
    ExecutionContext,
    ExecutionResult,
    GraphQLError,
    GraphQLSchema,
    TypeInfo,
    execute,
    execute_sync,
    parse,
    subscribe as _subscribe,
)
from graphql.execution import MiddlewareManager
from graphql.validation import specified_rules, validate
from graphql.validation.rules import ASTValidationRule

from .extensions import ExtensionManager
from .format_error import format_error
from .logger import log_error
from .types import (
    ErrorFormatter,
    ExtensionList,
    GraphQLResult,
    RootValue,
    SubscriptionResult,
    ValidationRules,
)
from .validation.introspection_disabled import IntrospectionDisabledRule


async def graphql(
    schema: GraphQLSchema,
    data: Any,
    *,
    context_value: Optional[Any] = None,
    root_value: Optional[RootValue] = None,
    debug: bool = False,
    introspection: bool = True,
    logger: Optional[str] = None,
    validation_rules: Optional[ValidationRules] = None,
    error_formatter: ErrorFormatter = format_error,
    middleware: Optional[MiddlewareManager] = None,
    extensions: Optional[ExtensionList] = None,
    **kwargs,
) -> GraphQLResult:
    extension_manager = ExtensionManager(extensions, context_value)

    with extension_manager.request():
        try:
            validate_data(data)
            query, variables, operation_name = (
                data["query"],
                data.get("variables"),
                data.get("operationName"),
            )

            document = parse_query(query)

            if callable(validation_rules):
                validation_rules = cast(
                    Optional[Collection[Type[ASTValidationRule]]],
                    validation_rules(context_value, document, data),
                )

            validation_errors = validate_query(
                schema, document, validation_rules, enable_introspection=introspection
            )
            if validation_errors:
                return handle_graphql_errors(
                    validation_errors,
                    logger=logger,
                    error_formatter=error_formatter,
                    debug=debug,
                    extension_manager=extension_manager,
                )

            if callable(root_value):
                root_value = root_value(context_value, document)
                if isawaitable(root_value):
                    root_value = await root_value

            result = execute(
                schema,
                document,
                root_value=root_value,
                context_value=context_value,
                variable_values=variables,
                operation_name=operation_name,
                execution_context_class=ExecutionContext,
                middleware=extension_manager.as_middleware_manager(middleware),
                **kwargs,
            )

            if isawaitable(result):
                result = await cast(Awaitable[ExecutionResult], result)
        except GraphQLError as error:
            return handle_graphql_errors(
                [error],
                logger=logger,
                error_formatter=error_formatter,
                debug=debug,
                extension_manager=extension_manager,
            )
        else:
            return handle_query_result(
                result,
                logger=logger,
                error_formatter=error_formatter,
                debug=debug,
                extension_manager=extension_manager,
            )


def graphql_sync(
    schema: GraphQLSchema,
    data: Any,
    *,
    context_value: Optional[Any] = None,
    root_value: Optional[RootValue] = None,
    debug: bool = False,
    introspection: bool = True,
    logger: Optional[str] = None,
    validation_rules: Optional[ValidationRules] = None,
    error_formatter: ErrorFormatter = format_error,
    middleware: Optional[MiddlewareManager] = None,
    extensions: Optional[ExtensionList] = None,
    **kwargs,
) -> GraphQLResult:
    extension_manager = ExtensionManager(extensions, context_value)

    with extension_manager.request():
        try:
            validate_data(data)
            query, variables, operation_name = (
                data["query"],
                data.get("variables"),
                data.get("operationName"),
            )

            document = parse_query(query)

            if callable(validation_rules):
                validation_rules = cast(
                    Optional[Collection[Type[ASTValidationRule]]],
                    validation_rules(context_value, document, data),
                )

            validation_errors = validate_query(
                schema, document, validation_rules, enable_introspection=introspection
            )
            if validation_errors:
                return handle_graphql_errors(
                    validation_errors,
                    logger=logger,
                    error_formatter=error_formatter,
                    debug=debug,
                    extension_manager=extension_manager,
                )

            if callable(root_value):
                root_value = root_value(context_value, document)
                if isawaitable(root_value):
                    ensure_future(root_value).cancel()
                    raise RuntimeError(
                        "Root value resolver can't be asynchronous "
                        "in synchronous query executor."
                    )

            result = execute_sync(
                schema,
                document,
                root_value=root_value,
                context_value=context_value,
                variable_values=variables,
                operation_name=operation_name,
                execution_context_class=ExecutionContext,
                middleware=extension_manager.as_middleware_manager(middleware),
                **kwargs,
            )

            if isawaitable(result):
                ensure_future(cast(Awaitable[ExecutionResult], result)).cancel()
                raise RuntimeError(
                    "GraphQL execution failed to complete synchronously."
                )
        except GraphQLError as error:
            return handle_graphql_errors(
                [error],
                logger=logger,
                error_formatter=error_formatter,
                debug=debug,
                extension_manager=extension_manager,
            )
        else:
            return handle_query_result(
                result,
                logger=logger,
                error_formatter=error_formatter,
                debug=debug,
                extension_manager=extension_manager,
            )


async def subscribe(
    schema: GraphQLSchema,
    data: Any,
    *,
    context_value: Optional[Any] = None,
    root_value: Optional[RootValue] = None,
    debug: bool = False,
    introspection: bool = True,
    logger: Optional[str] = None,
    validation_rules: Optional[ValidationRules] = None,
    error_formatter: ErrorFormatter = format_error,
    **kwargs,
) -> SubscriptionResult:
    try:
        validate_data(data)
        query, variables, operation_name = (
            data["query"],
            data.get("variables"),
            data.get("operationName"),
        )

        document = parse_query(query)

        if callable(validation_rules):
            validation_rules = cast(
                Optional[Collection[Type[ASTValidationRule]]],
                validation_rules(context_value, document, data),
            )

        validation_errors = validate_query(
            schema, document, validation_rules, enable_introspection=introspection
        )
        if validation_errors:
            for error_ in validation_errors:  # mypy issue #5080
                log_error(error_, logger)
            return (
                False,
                [error_formatter(error, debug) for error in validation_errors],
            )

        if callable(root_value):
            root_value = root_value(context_value, document)
            if isawaitable(root_value):
                root_value = await root_value

        result = await _subscribe(
            schema,
            document,
            root_value=root_value,
            context_value=context_value,
            variable_values=variables,
            operation_name=operation_name,
            **kwargs,
        )
    except GraphQLError as error:
        log_error(error, logger)
        return False, [error_formatter(error, debug)]
    else:
        if isinstance(result, ExecutionResult):
            errors = cast(List[GraphQLError], result.errors)
            for error_ in errors:  # mypy issue #5080
                log_error(error_, logger)
            return False, [error_formatter(error, debug) for error in errors]
        return True, cast(AsyncGenerator, result)


def handle_query_result(
    result, *, logger, error_formatter, debug, extension_manager=None
) -> GraphQLResult:
    response = {"data": result.data}
    if result.errors:
        for error in result.errors:
            log_error(error, logger)
        response["errors"] = [error_formatter(error, debug) for error in result.errors]

    if extension_manager:
        if result.errors:
            extension_manager.has_errors(result.errors)
        add_extensions_to_response(extension_manager, response)
    return True, response


def handle_graphql_errors(
    errors: Sequence[GraphQLError],
    *,
    logger,
    error_formatter,
    debug,
    extension_manager=None,
) -> GraphQLResult:
    for error in errors:
        log_error(error, logger)
    response = {"errors": [error_formatter(error, debug) for error in errors]}
    if extension_manager:
        extension_manager.has_errors(errors)
        add_extensions_to_response(extension_manager, response)
    return False, response


def parse_query(query):
    try:
        return parse(query)
    except GraphQLError as error:
        raise error
    except Exception as error:
        raise GraphQLError(str(error), original_error=error) from error


def add_extensions_to_response(extension_manager: ExtensionManager, response: dict):
    formatted_extensions = extension_manager.format()
    if formatted_extensions:
        if "extensions" in response:
            response["extensions"].update(formatted_extensions)
        else:
            response["extensions"] = formatted_extensions


def validate_query(
    schema: GraphQLSchema,
    document_ast: DocumentNode,
    rules: Optional[Collection[Type[ASTValidationRule]]] = None,
    max_errors: Optional[int] = None,
    type_info: Optional[TypeInfo] = None,
    enable_introspection: bool = True,
) -> List[GraphQLError]:
    if not enable_introspection:
        rules = (
            tuple(rules) + (IntrospectionDisabledRule,)
            if rules is not None
            else (IntrospectionDisabledRule,)
        )
    if rules:
        # run validation against rules from spec and custom rules
        supplemented_rules = specified_rules + tuple(rules)
        return validate(
            schema,
            document_ast,
            rules=supplemented_rules,
            max_errors=max_errors,
            type_info=type_info,
        )
    # run validation using spec rules only
    return validate(schema, document_ast, rules=specified_rules, type_info=type_info)


def validate_data(data: Optional[dict]) -> None:
    if not isinstance(data, dict):
        raise GraphQLError("Operation data should be a JSON object")
    validate_query_body(data.get("query"))
    validate_variables(data.get("variables"))
    validate_operation_name(data.get("operationName"))


def validate_query_body(query) -> None:
    if not query or not isinstance(query, str):
        raise GraphQLError("The query must be a string.")


def validate_variables(variables) -> None:
    if variables is not None and not isinstance(variables, dict):
        raise GraphQLError("Query variables must be a null or an object.")


def validate_operation_name(operation_name) -> None:
    if operation_name is not None and not isinstance(operation_name, str):
        raise GraphQLError('"%s" is not a valid operation name.' % operation_name)
