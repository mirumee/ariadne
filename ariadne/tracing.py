import datetime
import time
from inspect import isawaitable

from graphql import GraphQLResolveInfo, ResponsePath


def format_path(path: ResponsePath):
    elements = []
    while path:
        elements.append(path.key)
        path = path.prev
    return elements[::-1]


class TracingMiddleware:
    def __init__(self):
        self.start_date = datetime.datetime.utcnow()
        self.start_timestamp = time.perf_counter_ns()
        self.parsing_start_timestamp = self.start_timestamp
        self.parsing_end_timestamp = self.start_timestamp
        self.validation_start_timestamp = self.start_timestamp
        self.validation_end_timestamp = self.start_timestamp
        self.resolvers = []

    def start_validation(self):
        self.validation_start_timestamp = time.perf_counter_ns()

    def end_validation(self):
        self.validation_end_timestamp = time.perf_counter_ns()

    def should_trace(self, info: GraphQLResolveInfo):
        path = info.path
        while path:
            if isinstance(path.key, str) and path.key.startswith('__'):
                return False
            path = path.prev
        if info.parent_type.fields[info.field_name].resolve is None:
            return False
        return True

    async def resolve(self, next_, parent, info: GraphQLResolveInfo, **kwargs):
        if not self.should_trace(info):
            return next_(parent, info, **kwargs)

        start_timestamp = time.perf_counter_ns()
        record = {
            "path": format_path(info.path),
            "parentType": str(info.parent_type),
            "fieldName": info.field_name,
            "returnType": str(info.return_type),
            "startOffset": start_timestamp - self.start_timestamp,
        }
        self.resolvers.append(record)
        try:
            result = next_(parent, info, **kwargs)
            if isawaitable(result):
                result = await result
            return result
        finally:
            end_timestamp = time.perf_counter_ns()
            record["duration"] = end_timestamp - start_timestamp

    def extension_data(self):
        return {
            "tracing": {
                "version": 1,
                "startTime": self.start_date.strftime("%Y-%m-%dT%H:%M:%S.%fZ"),
                "endTime": datetime.datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%S.%fZ"),
                "duration": time.perf_counter_ns() - self.start_timestamp,
                "parsing": {
                    "startOffset": self.parsing_start_timestamp - self.start_timestamp,
                    "duration": self.parsing_end_timestamp
                    - self.parsing_start_timestamp,
                },
                "validation": {
                    "startOffset": self.validation_start_timestamp
                    - self.start_timestamp,
                    "duration": self.validation_end_timestamp
                    - self.validation_start_timestamp,
                },
                "execution": {"resolvers": self.resolvers},
            }
        }
