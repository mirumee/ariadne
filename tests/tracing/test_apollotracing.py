import pytest
from freezegun import freeze_time
from graphql import get_introspection_query

from ariadne import graphql, graphql_sync
from ariadne.contrib.tracing.apollotracing import ApolloTracingExtension


@pytest.mark.asyncio
async def test_apollotracing_extension_causes_no_errors_in_async_query_execution(
    async_schema,
):
    _, result = await graphql(
        async_schema, {"query": "{ status }"}, extensions=[ApolloTracingExtension]
    )
    assert result["data"] == {"status": True}


def test_apollotracing_extension_causes_no_errors_in_sync_query_execution(schema):
    _, result = graphql_sync(
        schema, {"query": "{ status }"}, extensions=[ApolloTracingExtension]
    )
    assert result["data"] == {"status": True}


@pytest.fixture
def freeze_microtime(mocker):
    mocker.patch(
        "ariadne.contrib.tracing.apollotracing.perf_counter_ns", return_value=2
    )


@freeze_time("2012-01-14 03:21:34")
@pytest.mark.asyncio
async def test_apollotracing_extension_adds_tracing_data_to_async_result_extensions(
    async_schema, freeze_microtime, snapshot  # pylint: disable=unused-argument
):
    _, result = await graphql(
        async_schema, {"query": "{ status }"}, extensions=[ApolloTracingExtension]
    )
    assert snapshot == result


@freeze_time("2012-01-14 03:21:34")
def test_apollotracing_extension_adds_tracing_data_to_sync_result_extensions(
    schema, freeze_microtime, snapshot  # pylint: disable=unused-argument
):
    _, result = graphql_sync(
        schema, {"query": "{ status }"}, extensions=[ApolloTracingExtension]
    )
    assert snapshot == result


@freeze_time("2012-01-14 03:21:34")
@pytest.mark.asyncio
async def test_apollotracing_extension_handles_exceptions_in_resolvers_in_async_context(
    async_schema, freeze_microtime, snapshot  # pylint: disable=unused-argument
):
    _, result = await graphql(
        async_schema, {"query": "{ testError }"}, extensions=[ApolloTracingExtension]
    )
    assert snapshot == result["data"]


@freeze_time("2012-01-14 03:21:34")
def test_apollotracing_extension_handles_exceptions_in_resolvers_in_sync_context(
    schema, freeze_microtime, snapshot  # pylint: disable=unused-argument
):
    _, result = graphql_sync(
        schema, {"query": "{ testError }"}, extensions=[ApolloTracingExtension]
    )
    assert snapshot == result["data"]


@pytest.mark.asyncio
async def test_apollotracing_extension_doesnt_break_introspection_in_async_context(
    async_schema,
):
    introspection_query = get_introspection_query(descriptions=True)
    _, result = await graphql(
        async_schema,
        {"query": introspection_query},
        extensions=[ApolloTracingExtension],
    )
    assert "errors" not in result


def test_apollotracing_extension_doesnt_break_introspection_in_sync_context(schema):
    introspection_query = get_introspection_query(descriptions=True)
    _, result = graphql_sync(
        schema, {"query": introspection_query}, extensions=[ApolloTracingExtension]
    )
    assert "errors" not in result
