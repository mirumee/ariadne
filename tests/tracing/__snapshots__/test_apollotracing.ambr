# serializer version: 1
# name: test_apollotracing_extension_adds_tracing_data_to_async_result_extensions
  dict({
    'data': dict({
      'status': True,
    }),
    'extensions': dict({
      'tracing': dict({
        'duration': 0,
        'endTime': '2012-01-14T03:21:34.000000Z',
        'execution': dict({
          'resolvers': list([
            dict({
              'duration': 0,
              'fieldName': 'status',
              'parentType': 'Query',
              'path': list([
                'status',
              ]),
              'returnType': 'Boolean',
              'startOffset': 0,
            }),
          ]),
        }),
        'startTime': '2012-01-14T03:21:34.000000Z',
        'version': 1,
      }),
    }),
  })
# ---
# name: test_apollotracing_extension_adds_tracing_data_to_sync_result_extensions
  dict({
    'data': dict({
      'status': True,
    }),
    'extensions': dict({
      'tracing': dict({
        'duration': 0,
        'endTime': '2012-01-14T03:21:34.000000Z',
        'execution': dict({
          'resolvers': list([
            dict({
              'duration': 0,
              'fieldName': 'status',
              'parentType': 'Query',
              'path': list([
                'status',
              ]),
              'returnType': 'Boolean',
              'startOffset': 0,
            }),
          ]),
        }),
        'startTime': '2012-01-14T03:21:34.000000Z',
        'version': 1,
      }),
    }),
  })
# ---
# name: test_apollotracing_extension_handles_exceptions_in_resolvers_in_async_context
  dict({
    'testError': None,
  })
# ---
# name: test_apollotracing_extension_handles_exceptions_in_resolvers_in_sync_context
  dict({
    'testError': None,
  })
# ---
